File: engine/src/main/java/de/hanno/hpengine/engine/model/texture/CubeMap.java
package de.hanno.hpengine.engine.model.texture;

import de.hanno.hpengine.engine.graphics.GpuContext;
import de.hanno.hpengine.engine.graphics.renderer.constants.GlTextureTarget;
import org.jetbrains.annotations.NotNull;
import org.lwjgl.opengl.EXTTextureSRGB;
import org.lwjgl.opengl.GL11;

import java.io.Serializable;
import java.nio.ByteBuffer;
import java.nio.ByteOrder;
import java.util.List;

import static de.hanno.hpengine.engine.graphics.renderer.constants.GlTextureTarget.TEXTURE_CUBE_MAP;

public class CubeMap implements CubeTexture, Serializable {
	private static final long serialVersionUID = 1L;
	protected final TextureManager textureManager;
	private final String path;
	protected final GlTextureTarget target;

	public List<byte[]> dataList;
	protected final int width;
	protected final int height;
	protected final int minFilter;
	protected final int magFilter;
	protected int srcPixelFormat;
	protected int dstPixelFormat;
	protected long handle;
	protected int textureId;
	protected long lastUsedTimeStamp;

	public CubeMap(TextureManager textureManager, String path, GlTextureTarget target, int width, int height, int minFilter, int magFilter, int srcPixelFormat, int dstPixelFormat) {
		this.textureManager = textureManager;
		this.path = path;
		this.target = target;
		this.width = width;
		this.height = height;
		this.minFilter = minFilter;
		this.magFilter = magFilter;
		this.srcPixelFormat = srcPixelFormat;
		this.dstPixelFormat = dstPixelFormat;
	}
	
	public ByteBuffer buffer(ByteBuffer buffer, byte[] values) {
		buffer.order(ByteOrder.nativeOrder());
		buffer.put(values, 0, values.length);
		buffer.flip();
		return buffer;
	}
	
	public void load(int cubemapFace, ByteBuffer buffer) {
        GL11.glTexImage2D(cubemapFace,
                0, 
				EXTTextureSRGB.GL_COMPRESSED_SRGB_ALPHA_S3TC_DXT5_EXT,
				(getWidth()/4),
				(getHeight()/3),
                0, 
                srcPixelFormat, 
                GL11.GL_UNSIGNED_BYTE, 
                buffer);
	}

	public void setData(List<byte[]> byteArrays) {
		dataList = byteArrays;
	}
	
	@Override
	public String toString() {
		return "(Cubemap)" + path;
	}

	public void bind(GpuContext gpuContext, int unit) {
		gpuContext.bindTexture(unit, TEXTURE_CUBE_MAP, textureId);
	}

	@Override
	public int getTextureId() {
		return textureId;
	}

	@NotNull
	@Override
	public GlTextureTarget getTarget() {
		return target;
	}

	@Override
	public void setTarget(@NotNull GlTextureTarget glTextureTarget) {
		// TODO: Remove this
		throw new IllegalStateException("");
	}

	@Override
	public long getHandle() {
		return handle;
	}

	@Override
	public int getWidth() {
		return width;
	}

	@Override
	public int getHeight() {
		return height;
	}

	@Override
	public long getLastUsedTimeStamp() {
		return lastUsedTimeStamp;
	}

	@Override
	public int getMinFilter() {
		return minFilter;
	}

	@Override
	public int getMagFilter() {
		return magFilter;
	}

	@Override
	public void unload() {}

	void bind() {
		textureManager.getGpuContext().bindTexture(target, textureId);
	}

	public void bind(int textureUnitIndex) {
		textureManager.getGpuContext().bindTexture(textureUnitIndex, target, textureId);
	}

	public void setHandle(long handle) {
		this.handle = handle;
	}

	@Override
	public void setUsedNow() {
//		TODO Implement me
	}

	@Override
	public List<? extends byte[]> getData() {
		return dataList;
	}
}

File: engine/src/main/java/de/hanno/hpengine/engine/model/texture/CubeMap.kt
package de.hanno.hpengine.engine.model.texture

import de.hanno.hpengine.engine.graphics.renderer.constants.GlTextureTarget.TEXTURE_CUBE_MAP
import de.hanno.hpengine.engine.graphics.renderer.constants.TextureFilterConfig
import de.hanno.hpengine.util.Util
import org.lwjgl.opengl.EXTTextureCompressionS3TC
import org.lwjgl.opengl.EXTTextureSRGB
import org.lwjgl.opengl.GL11
import org.lwjgl.opengl.GL11.glTexParameteri
import org.lwjgl.opengl.GL12
import java.io.Serializable
import java.nio.ByteBuffer
import java.nio.ByteOrder

interface CubeTexture : Texture<TextureDimension3D>

open class CubeMap(protected val textureManager: TextureManager,
                   private val path: String,
                   override val dimension: TextureDimension3D,
                   override val textureFilterConfig: TextureFilterConfig = TextureFilterConfig(),
                   protected var srcPixelFormat: Int,
                   override val id: Int,
                   private val data: MutableList<ByteArray>) : CubeTexture, Serializable {
    private val srgba = true // TODO: Make this configurable
    override var handle: Long = -1
    override val target = TEXTURE_CUBE_MAP
    override val wrapMode = GL12.GL_CLAMP_TO_EDGE // TODO: Make this configurable
    override val internalFormat: Int = if (srgba) EXTTextureSRGB.GL_COMPRESSED_SRGB_ALPHA_S3TC_DXT5_EXT else EXTTextureCompressionS3TC.GL_COMPRESSED_RGBA_S3TC_DXT5_EXT
    override var uploadState = UploadState.NOT_UPLOADED

    init {
        textureManager.gpuContext.bindTexture(this)
        setupTextureParameters()
    }

    fun setupTextureParameters() = textureManager.gpuContext.execute("setupTextureParameters") {
        glTexParameteri(target.glTarget, GL11.GL_TEXTURE_MIN_FILTER, textureFilterConfig.minFilter.glValue)
        glTexParameteri(target.glTarget, GL11.GL_TEXTURE_MAG_FILTER, textureFilterConfig.magFilter.glValue)
        glTexParameteri(target.glTarget, GL12.GL_TEXTURE_WRAP_R, wrapMode)
        glTexParameteri(target.glTarget, GL11.GL_TEXTURE_WRAP_S, wrapMode)
        glTexParameteri(target.glTarget, GL11.GL_TEXTURE_WRAP_T, wrapMode)
        glTexParameteri(target.glTarget, GL12.GL_TEXTURE_BASE_LEVEL, 0)
        glTexParameteri(target.glTarget, GL12.GL_TEXTURE_MAX_LEVEL, Util.calculateMipMapCount(Math.max(dimension.width, dimension.height)))
    }

    fun load(cubeMapFace: Int, buffer: ByteBuffer) {
        GL11.glTexImage2D(cubeMapFace,
                0,
                internalFormat,
                dimension.width / 4,
                dimension.height / 3,
                0,
                srcPixelFormat,
                GL11.GL_UNSIGNED_BYTE,
                buffer)
    }

    override fun toString(): String {
        return "(Cubemap)$path"
    }

    fun getData(): List<ByteArray> {
        return data
    }

    companion object {
        private const val serialVersionUID = 1L

        @JvmStatic
        fun buffer(buffer: ByteBuffer, values: ByteArray): ByteBuffer {
            buffer.order(ByteOrder.nativeOrder())
            buffer.put(values, 0, values.size)
            buffer.flip()
            return buffer
        }
    }
}
